File: src/ui/redirect.rs
//! The loopback listener the sign-in redirect lands on.
//!
//! Battle.net rejects custom URI schemes — the redirect must begin with `http`
//! or `https` — and it matches the registered string exactly, with no RFC 8252
//! loopback port wildcard. So a real HTTP server on a fixed port on 127.0.0.1 is
//! not a shortcut, it is the only supported shape for a desktop client.
//!
//! `soup::Server` is what serves it. libsoup is already here for the API calls,
//! it is already in `org.gnome.Sdk`, and it runs on the GLib main loop — so the
//! listener costs no thread and no second dependency.

use std::cell::RefCell;
use std::rc::Rc;

use gtk::glib;
use soup::prelude::*;

use crate::model::source::blizzard::oauth;
use crate::model::source::Reason;

/// A one-shot listener for the authorization code.
pub struct Redirect {
    server: soup::Server,
}

impl Redirect {
    /// Start listening, and call `deliver` when the browser comes back.
    ///
    /// `state` is what was sent to the authorize endpoint. A redirect carrying
    /// anything else did not come from the flow we started and is refused —
    /// with no PKCE verifier to fall back on, this check is the only thing
    /// standing between the flow and a code somebody else planted.
    pub fn listen<F>(state: &str, deliver: F) -> Result<Redirect, glib::Error>
    where
        F: Fn(Result<String, Reason>) + 'static,
    {
        let server = soup::Server::builder().server_header("Armory").build();

        let expected = state.to_string();
        // The browser will happily request /favicon.ico alongside the callback,
        // and a retried or refreshed redirect delivers the same code twice.
        // Answering only once keeps the flow from being driven backwards.
        let spent = Rc::new(RefCell::new(false));

        server.add_handler(
            Some("/callback"),
            glib::clone!(
                #[strong]
                expected,
                #[strong]
                spent,
                move |_server: &soup::Server,
                      message: &soup::ServerMessage,
                      _path: &str,
                      _query: std::collections::HashMap<&str, &str>| {
                    // The parsed query map libsoup hands over is borrowed and
                    // already percent-decoded; the raw string is read back off
                    // the URI instead so that `oauth::parse_callback` — which is
                    // tested against real redirects — owns the decoding.
                    let uri = message.uri().map(|uri| uri.to_string()).unwrap_or_default();
                    let query = uri.split_once('?').map(|(_, query)| query).unwrap_or("");

                    if std::mem::replace(&mut *spent.borrow_mut(), true) {
                        respond(message, Landing::AlreadyDone);
                        return;
                    }

                    match oauth::parse_callback(query) {
                        Ok((code, state)) if state == expected => {
                            respond(message, Landing::SignedIn);
                            deliver(Ok(code));
                        }
                        Ok(_) => {
                            respond(message, Landing::Mismatch);
                            deliver(Err(Reason::Unauthorised(
                                "the sign-in did not come back from the request Armory started"
                                    .into(),
                            )));
                        }
                        Err(reason) => {
                            respond(message, Landing::Refused);
                            deliver(Err(reason));
                        }
                    }
                }
            ),
        );

        server.listen_local(
            oauth::REDIRECT_PORT as u32,
            soup::ServerListenOptions::IPV4_ONLY,
        )?;

        Ok(Redirect { server })
    }
}

impl Drop for Redirect {
    fn drop(&mut self) {
        // The port is fixed and registered, so leaving it bound would make the
        // next sign-in attempt fail with something that reads like a Blizzard
        // problem.
        soup::prelude::ServerExt::disconnect(&self.server);
    }
}

fn respond(message: &soup::ServerMessage, landing: Landing) {
    message.set_status(200, None);
    message.set_response(
        Some("text/html; charset=utf-8"),
        soup::MemoryUse::Copy,
        page(landing).as_bytes(),
    );
}

/// What the browser is told, in each of the four cases it can land in.
#[derive(Clone, Copy)]
enum Landing {
    SignedIn,
    Mismatch,
    Refused,
    AlreadyDone,
}

impl Landing {
    fn heading(self) -> &'static str {
        match self {
            Landing::SignedIn => "Signed in",
            Landing::Mismatch => "That did not come from Armory",
            Landing::Refused => "Sign-in cancelled",
            Landing::AlreadyDone => "Already done",
        }
    }

    fn body(self) -> &'static str {
        match self {
            Landing::SignedIn => "You can close this tab and go back to Armory.",
            Landing::Mismatch => {
                "The sign-in response did not match the one Armory started, so it has \
                 been ignored and nothing has been saved. Start again from Armory."
            }
            Landing::Refused => {
                "Nothing has been saved. You can try again from Armory whenever you like."
            }
            Landing::AlreadyDone => "You can close this tab.",
        }
    }
}

/// Render one of them.
///
/// No external anything: no fonts, no scripts, no images. This is served to a
/// browser by a process that is not a web server and should not start behaving
/// like one — and a page that fetches something is a page that can hang on a
/// machine with no route out.
fn page(landing: Landing) -> String {
    format!(
        "<!doctype html><meta charset=utf-8><title>Armory</title>\
         <style>body{{font-family:system-ui,sans-serif;margin:4rem auto;max-width:30rem;\
         padding:0 1rem;line-height:1.5;color:#241f31;background:#fafafa}}\
         @media(prefers-color-scheme:dark){{body{{color:#deddda;background:#1d1d20}}}}\
         h1{{font-size:1.3rem}}</style><h1>{}</h1><p>{}</p>",
        landing.heading(),
        landing.body()
    )
}

File: dotnet/Armory.Client/Blizzard/Redirect.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using Armory.Blizzard;

namespace Armory.Client.Blizzard;

/// <summary>
/// The loopback listener the sign-in redirect lands on.
///
/// Battle.net rejects custom URI schemes and matches the registered redirect
/// exactly, with no loopback port wildcard, so a real HTTP server on a fixed
/// port on 127.0.0.1 is the only supported shape for a desktop client. It is
/// a raw socket rather than <c>HttpListener</c>: that one needs a URL
/// reservation on Windows for anything but an administrator, and a sign-in
/// that fails for want of a <c>netsh</c> command reads like a Blizzard
/// problem.
///
/// One shot. The browser will happily request <c>/favicon.ico</c> alongside
/// the callback, and a retried or refreshed redirect delivers the same code
/// twice; answering only once keeps the flow from being driven backwards.
/// Disposing unbinds the port, and the port is fixed and registered, so it
/// must not leak.
/// </summary>
public sealed class Redirect : IDisposable
{
    private readonly TcpListener listener;
    private readonly string expected;
    private readonly Action<Result<string, Reason>> deliver;
    private readonly CancellationTokenSource stopping = new();
    private bool spent;

    private Redirect(TcpListener listener, string expected, Action<Result<string, Reason>> deliver)
    {
        this.listener = listener;
        this.expected = expected;
        this.deliver = deliver;
    }

    /// <summary>The port this one is bound to. The registered one unless a test asked otherwise.</summary>
    public int Port { get; private init; }

    /// <summary>
    /// Start listening, and call <paramref name="deliver"/> when the browser
    /// comes back. <paramref name="state"/> is what was sent to the authorize
    /// endpoint; a redirect carrying anything else did not come from the flow
    /// we started and is refused. With no PKCE verifier to fall back on, this
    /// check is the only thing between the flow and a code somebody planted.
    /// </summary>
    public static Result<Redirect, string> Listen(string state, Action<Result<string, Reason>> deliver, int port = OAuth.RedirectPort)
    {
        var listener = new TcpListener(IPAddress.Loopback, port);
        try
        {
            listener.Start();
        }
        catch (SocketException error)
        {
            return Result<Redirect, string>.Err(error.Message);
        }
        var redirect = new Redirect(listener, state, deliver) { Port = ((IPEndPoint)listener.LocalEndpoint).Port };
        _ = redirect.Serve();
        return Result<Redirect, string>.Ok(redirect);
    }

    private async Task Serve()
    {
        while (!stopping.IsCancellationRequested)
        {
            TcpClient client;
            try
            {
                client = await listener.AcceptTcpClientAsync(stopping.Token);
            }
            catch (OperationCanceledException)
            {
                return;
            }
            catch (SocketException)
            {
                return;
            }
            catch (ObjectDisposedException)
            {
                return;
            }
            using (client)
            {
                await Answer(client);
            }
        }
    }

    private async Task Answer(TcpClient client)
    {
        string head;
        try
        {
            client.ReceiveTimeout = 5_000;
            head = await ReadHead(client.GetStream(), stopping.Token);
        }
        catch (Exception error) when (error is IOException or SocketException or OperationCanceledException or ObjectDisposedException)
        {
            return;
        }

        // The request line: `GET /callback?code=...&state=... HTTP/1.1`.
        var line = head.Split("\r\n", 2)[0];
        var parts = line.Split(' ');
        var target = parts.Length >= 2 ? parts[1] : "";
        var path = target.Split('?', 2)[0];
        var query = target.Contains('?', StringComparison.Ordinal) ? target.Split('?', 2)[1] : "";

        Landing landing;
        Result<string, Reason>? outcome = null;
        if (path != "/callback")
        {
            landing = Landing.NotHere;
        }
        else if (spent)
        {
            landing = Landing.AlreadyDone;
        }
        else
        {
            spent = true;
            var parsed = OAuth.ParseCallback(query);
            if (!parsed.IsOk)
            {
                landing = Landing.Refused;
                outcome = Result<string, Reason>.Err(parsed.Error);
            }
            else if (parsed.Value.State != expected)
            {
                landing = Landing.Mismatch;
                outcome = Result<string, Reason>.Err(new Reason.Unauthorised("the sign-in did not come back from the request Armory started"));
            }
            else
            {
                landing = Landing.SignedIn;
                outcome = Result<string, Reason>.Ok(parsed.Value.Code);
            }
        }

        var body = Encoding.UTF8.GetBytes(Page(landing));
        var status = landing == Landing.NotHere ? "404 Not Found" : "200 OK";
        var response = Encoding.ASCII.GetBytes($"HTTP/1.1 {status}\r\nServer: Armory\r\nContent-Type: text/html; charset=utf-8\r\nContent-Length: {body.Length}\r\nConnection: close\r\n\r\n");
        try
        {
            var stream = client.GetStream();
            await stream.WriteAsync(response, stopping.Token);
            await stream.WriteAsync(body, stopping.Token);
            await stream.FlushAsync(stopping.Token);
        }
        catch (Exception error) when (error is IOException or SocketException or OperationCanceledException or ObjectDisposedException)
        {
            // The browser went away; the code still counts.
        }

        if (outcome is { } delivered)
        {
            deliver(delivered);
        }
    }

    /// <summary>The request head, up to the blank line. The body, if any, is not wanted.</summary>
    private static async Task<string> ReadHead(NetworkStream stream, CancellationToken cancellation)
    {
        var buffer = new byte[8192];
        var held = new MemoryStream();
        while (held.Length < buffer.Length)
        {
            var read = await stream.ReadAsync(buffer.AsMemory(0, buffer.Length - (int)held.Length), cancellation);
            if (read == 0)
            {
                break;
            }
            held.Write(buffer, 0, read);
            var text = Encoding.ASCII.GetString(held.GetBuffer(), 0, (int)held.Length);
            var end = text.IndexOf("\r\n\r\n", StringComparison.Ordinal);
            if (end >= 0)
            {
                return text[..end];
            }
        }
        return Encoding.ASCII.GetString(held.GetBuffer(), 0, (int)held.Length);
    }

    public void Dispose()
    {
        stopping.Cancel();
        listener.Stop();
        stopping.Dispose();
    }

    /// <summary>What the browser is told, in each of the cases it can land in.</summary>
    internal enum Landing
    {
        SignedIn,
        Mismatch,
        Refused,
        AlreadyDone,
        NotHere,
    }

    internal static string Heading(Landing landing) => landing switch
    {
        Landing.SignedIn => "Signed in",
        Landing.Mismatch => "That did not come from Armory",
        Landing.Refused => "Sign-in cancelled",
        Landing.AlreadyDone => "Already done",
        _ => "Nothing here",
    };

    internal static string Body(Landing landing) => landing switch
    {
        Landing.SignedIn => "You can close this tab and go back to Armory.",
        Landing.Mismatch => "The sign-in response did not match the one Armory started, so it has been ignored and nothing has been saved. Start again from Armory.",
        Landing.Refused => "Nothing has been saved. You can try again from Armory whenever you like.",
        Landing.AlreadyDone => "You can close this tab.",
        _ => "Armory is listening for a Battle.net sign-in, and this is not one.",
    };

    /// <summary>
    /// Render one of them. No external anything: no fonts, no scripts, no
    /// images. This is served to a browser by a process that is not a web
    /// server, and a page that fetches something can hang on a machine with
    /// no route out.
    /// </summary>
    internal static string Page(Landing landing) =>
        "<!doctype html><meta charset=utf-8><title>Armory</title>" +
        "<style>body{font-family:system-ui,sans-serif;margin:4rem auto;max-width:30rem;padding:0 1rem;line-height:1.5;color:#241f31;background:#fafafa}" +
        "@media(prefers-color-scheme:dark){body{color:#deddda;background:#1d1d20}}h1{font-size:1.3rem}</style>" +
        $"<h1>{Heading(landing)}</h1><p>{Body(landing)}</p>";
}
